### PR TITLE
Metaphlan4 update: add --index

### PR DESCRIPTION
**Pull request by LilyAnderssonLee**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/metaphlan/metaphlan/main.nf [hunk 1]
```diff
@@ -23,12 +23,13 @@ process METAPHLAN_METAPHLAN {
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    def input_type  = ("$input".endsWith(".fastq.gz") || "$input".endsWith(".fq.gz")) ? "--input_type fastq" :  ("$input".contains(".fasta")) ? "--input_type fasta" : ("$input".endsWith(".bowtie2out.txt")) ? "--input_type bowtie2out" : "--input_type sam"
+    def input_type  = ("$input".endsWith(".fastq.gz") || "$input".endsWith(".fq.gz")) ? "--input_type fastq" :  ("$input".contains(".fasta")|| "$input".endsWith(".fa.gz")) ? "--input_type fasta" : ("$input".endsWith(".bowtie2out.txt")) ? "--input_type bowtie2out" : "--input_type sam"
```
Comment (jfy133) [inline modules/nf-core/metaphlan/metaphlan/main.nf:26]: Are `fasta` and `.fa.gz` the only file suffixes allowed for MetaPhlAn4 for fasta input??

File: modules/nf-core/metaphlan/metaphlan/main.nf [hunk 2]
```diff
@@ -23,12 +23,13 @@ process METAPHLAN_METAPHLAN {
...
     def input_data  = ("$input_type".contains("fastq")) && !meta.single_end ? "${input[0]},${input[1]}" : "$input"
     def bowtie2_out = "$input_type" == "--input_type bowtie2out" || "$input_type" == "--input_type sam" ? '' : "--bowtie2out ${prefix}.bowtie2out.txt"
 
     """
     BT2_DB=`find -L "${metaphlan_db_latest}" -name "*rev.1.bt2l" -exec dirname {} \\;`
+    BT2_DB_INDEX=`find -L ${metaphlan_db_latest} -name "*.rev.1.bt2l" | sed 's/\\.rev.1.bt2l\$//' | sed 's/.*\\///'`
```
Comment (jfy133) [inline modules/nf-core/metaphlan/metaphlan/main.nf:32]: What are the differences between these two/

**Review by jfy133 — APPROVED** [review]
I guess looks conceptually OK but I'm a bit wary about the hardcoded specific FASTA indicies, and I don't understand what why bowh bowtie2db and index are needed?

File: modules/nf-core/metaphlan/metaphlan/main.nf (hunk 1 above)
Comment (LilyAnderssonLee) [reply modules/nf-core/metaphlan/metaphlan/main.nf:26]: Good point! It also includes `fasta.gz`

**Comment by LilyAnderssonLee** [comment]
> I guess looks conceptually OK but I'm a bit wary about the hardcoded specific FASTA indicies, and I don't understand what why bowh bowtie2db and index are needed?

To be honest, I'm not entirely sure why both bowtie2db and index are needed. However, the process will automatically download the database from the web without requiring --index. According to the MetaPhlAn4 manual, adding --index can help skip checking for the latest database version and avoid downloading it. 

MetaPhlAn3 doesn't seem to have this issue even without using --index and I haven't encountered this problem before.

File: modules/nf-core/metaphlan/metaphlan/main.nf (hunk 1 above)
Comment (jfy133) [reply modules/nf-core/metaphlan/metaphlan/main.nf:26]: OK, maybe there could be a way to simplify this... people could also supply `fna` or `fna.gz` as welll...

**Comment by jfy133** [comment]
> > I guess looks conceptually OK but I'm a bit wary about the hardcoded specific FASTA indicies, and I don't understand what why bowh bowtie2db and index are needed?
> 
> To be honest, I'm not entirely sure why both bowtie2db and index are needed. However, the process will automatically download the database from the web without requiring --index. According to the MetaPhlAn4 manual, adding --index can help skip checking for the latest database version and avoid downloading it.
> 
> MetaPhlAn3 doesn't seem to have this issue even without using --index and I haven't encountered this problem before.

Uh weird... but OK, then no harm in it! Then that comment is resolved :)

File: modules/nf-core/metaphlan/metaphlan/main.nf (hunk 1 above)
Comment (jfy133) [reply modules/nf-core/metaphlan/metaphlan/main.nf:26]: Yes much nicer! I guess it just missing the `gz` variants (if supported by the tool?)

Comment (LilyAnderssonLee) [reply modules/nf-core/metaphlan/metaphlan/main.nf:26]: I removed the $ at the end of the searching pattern now, so it should include gz variabts.

**Pull request merged**